---
title: "RNAseqAnalysis"
author: "Mhyles Hintural"
date: "26/08/2020"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set-up

```{r}
library(plyr)
library(reshape2)
library(dplyr)
library(tidyverse)
library(magrittr)
library(edgeR)
library(RColorBrewer)
library(limma)
library(ggplot2)
library(readxl)
library(pheatmap)
library(here)
library(xtable)
library(biomaRt)
library(org.Hs.eg.db)
library(factoextra)
```

# Data packaging
## Importing count data

```{r ImportCounts}
 
# Read in bam counts
termPlacentaCounts <- read.table(here("rawData/bam_counts.table.tsv"),
                                 sep = '\t',
                                 header = TRUE,
                                 row.names = "Geneid") %>%
  t() %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  filter(str_detect(samplename, "^S")) %>% 
  tibble::column_to_rownames("samplename") %>% 
  t()

 
# clean up the column headers and remove the file extension names
colnames(termPlacentaCounts) <- gsub("_Aligned.sortedByCoord.out.bam", "", colnames(termPlacentaCounts))
 
 
names <- colnames(termPlacentaCounts) %>%
  as.data.frame() %>%
  separate(., col = ., into = c("cohort", "samplenumber", "tissue", "lane"), sep = "_") %>%
  mutate(., samplename = paste(cohort, samplenumber, sep = "")) %>%
  dplyr::select(., samplename)
 
# convert back to a vector so I can overwrite the colnames in the counts table
names <- names[,]
 
# replace the old sample names with the new names
colnames(termPlacentaCounts) <- names

termPlacentaCounts <- termPlacentaCounts[,order(colnames(termPlacentaCounts), decreasing=FALSE)]

```

## Remove mRNA with zero counts in all samples

```{r filterZero}

# check to establish if any mRNA have no expression across all samples
table(rowSums(termPlacentaCounts == 0) == ncol(termPlacentaCounts))
# remove mRNAs with zero counts in all samples
termPlacentaCounts <- termPlacentaCounts[ rowSums(termPlacentaCounts)!=0, ]

```

## Importing meta-data

```{r ImportMetadata}

# Read in metadata

simpleMetadata <- read_excel(here("cleanData/simpleMetadata_matched.xlsx")) %>% 
  dplyr::arrange(., samplename) %>% 
  dplyr::filter(str_detect(samplename, "^S"))

```

## Creating a DGEList

```{r DGEList}

# new data frame of gene names for DGEList object
genes <- as.character(rownames(termPlacentaCounts)) %>%
  as.data.frame() %>%
  set_colnames("SYMBOL")

# new DGEList object for differential expression analysis
DGEList_placenta <- DGEList(counts = termPlacentaCounts,
                            samples = simpleMetadata,
                            genes = genes)

```

# Data pre-processing
## Visualise distribution of expression prior to filtering

```{r pre-filtering}

# new dataframe of counts per million
cpm <- cpm(DGEList_placenta)
# new dataframe of log2 counts per million
lcpm <-cpm(DGEList_placenta, log=TRUE)

# plot the density of counts pre-filtering for low expression
nsamples <- ncol(DGEList_placenta)
col = colorRampPalette(brewer.pal(11,"Spectral"))(ncol(DGEList_placenta))
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.6), las=2,
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}

```

## Removing low expression genes

```{r filterLow}

# which genes are expressed at > 1 counts per million, in at least 6 samples
# we choose 6 samples because prelabour LSCS has 6 patients, which is our smallest group
keep.exprs <- rowSums(cpm >1) >= 6
keep.exprs %>% table()
# only keep the genes that met the criteria above
DGEList_placenta <- DGEList_placenta[keep.exprs,, keep.lib.sizes = FALSE]

```

- removing genes with low expression has reduced our gene count from 45312 to 15404 genes.

## Visualise distribution of expression after filtering

```{r post-filtering}

cpm_post <- cpm(DGEList_placenta)
lcpm_post <-cpm(DGEList_placenta, log=TRUE)
nsamples <- ncol(DGEList_placenta)
col = colorRampPalette(brewer.pal(11,"Spectral"))(ncol(DGEList_placenta))
par(mfrow=c(1,2))
plot(density(lcpm_post[,1]), col=col[1], lwd=2, ylim=c(0,0.60), las=2,
     main="", xlab="")
title(main="A. Filtered data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm_post[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}

# Plot the distribution of un-normalised counts
un_normalised_lcpm_post <- cpm(DGEList_placenta, log=TRUE)
boxplot(un_normalised_lcpm_post, las=2, col=col, main="")
title(main = "Placenta: Un-normalised data (RNA)", ylab = "Log-cpm") 

```

## Normalising gene expression distributions

``` {r Normalisation}

# DGEList_placenta <- calcNormFactors(x, method = "TMM")
# DGEList_placenta$samples$norm.factors
# x2 <- DGEList_placenta
# x2$samples$norm.factors <- 1
# x2$counts[,1] <- ceiling(x2$counts[,1]*0.05)
# x2$counts[,2] <- x2$counts[,2]*5
# par(mfrow=c(1,2))
# lcpm <- cpm(x2, log=TRUE)
# boxplot(lcpm, las=2, col=col, main="")
# title(main="A. Unnormalised data",ylab="Log-cpm")
# x2 <- calcNormFactors(x2)  
# x2$samples$norm.factors
# lcpm <- cpm(x2, log=TRUE)
# boxplot(lcpm, las=2, col=col, main="")
# title(main="B. Normalised data",ylab="Log-cpm")



```

## TMM normalisation of libraries

```{r normalisation}

# Distribution of un-normalised counts 
nsamples <- ncol(DGEList_placenta)
col = colorRampPalette(brewer.pal(11,"Spectral"))(ncol(DGEList_placenta))
# calculate normalisation factors and apply to the DGEList object
DGEList_placenta <- calcNormFactors(DGEList_placenta, method = "TMM")
# Plot the distribution of counts after normalisation
normalised_lcpm <- cpm(DGEList_placenta, log = TRUE)
boxplot(normalised_lcpm, las = 2, col = col, main = "")
title(main = "Placenta: Normalised data (RNA)", ylab = "Log-cpm")

```

# Exploratory plots
## Unsupervised clustering of samples

``` {r MDSplots}

lcpm <- cpm(DGEList_placenta, log=TRUE)
# col.group <- nsamples
# levels(col.group) <-  brewer.pal(nlevels(col.group), "Set1")
# col.group <- as.character(col.group)
# col.lane <- lane
# levels(col.lane) <-  brewer.pal(nlevels(col.lane), "Set2")
# col.lane <- as.character(col.lane)
# plotMDS(lcpm, labels=delivery_mode, col=col.group)
# title(main="A. Sample groups")
# plotMDS(lcpm, labels=lane, col=col.lane, dim=c(3,4))
# title(main="B. Sequencing lanes")

col.mode <- DGEList_placenta$samples$delivery_mode
levels(col.mode) <-  brewer.pal(nlevels(col.mode), "Set1")
col.mode <- as.character(col.mode)

plotMDS(lcpm)

# MDS filtered and normalised counts

MDS <- plotMDS(DGEList_placenta, main = "MDS Filtered and Normalised Counts")

# pull out the x and y
x <- MDS$x
y <- MDS$y

# plot with ggplot to observe delivery mode vs fetal sex

cbind(x,y) %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  full_join(., DGEList_placenta$samples, by = "samplename") %>% 
  ggplot() +
  geom_point(aes(x = x,
                 y = y,
                 colour = delivery_mode,
                 shape = fetal_sex),
             size = 6) +
  geom_text(aes(x = x,
                y = y,
                label = samplename,
                colour = delivery_mode), vjust = -1) +
  labs(title = "MDS plot filtered and normalised counts",
       x = "Dim 1",
       y = "Dim 2") +
  theme_bw()

# plot with ggplot to observe delivery mode vs patient outcome

cbind(x,y) %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  full_join(., DGEList_placenta$samples, by = "samplename") %>% 
  ggplot() +
  geom_point(aes(x = x,
                 y = y,
                 colour = delivery_mode,
                 shape = PatientOutcome),
             size = 6) +
  geom_text(aes(x = x,
                y = y,
                label = samplename,
                colour = delivery_mode), vjust = -1) +
  labs(title = "MDS plot filtered and normalised counts",
       x = "Dim 1",
       y = "Dim 2") +
  theme_bw()

# plot with ggplot to observe delivery mode vs maternal BMI

cbind(x,y) %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  full_join(., DGEList_placenta$samples, by = "samplename") %>% 
  ggplot() +
  geom_point(aes(x = x,
                 y = y,
                 colour = BMI,
                 shape = delivery_mode),
             size = 6) +
  geom_text(aes(x = x,
                y = y,
                label = samplename,
                colour = BMI), vjust = -1) +
  labs(title = "MDS plot filtered and normalised counts",
       x = "Dim 1",
       y = "Dim 2") +
  theme_bw()

# plot with ggplot to observe delivery mode vs birth weight

cbind(x,y) %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  full_join(., DGEList_placenta$samples, by = "samplename") %>% 
  ggplot() +
  geom_point(aes(x = x,
                 y = y,
                 colour = birthwgt,
                 shape = delivery_mode),
             size = 6) +
  geom_text(aes(x = x,
                y = y,
                label = samplename,
                colour = birthwgt), vjust = -1) +
  labs(title = "MDS plot filtered and normalised counts",
       x = "Dim 1",
       y = "Dim 2") +
  theme_bw()

# plot with ggplot to observe delivery mode vs maternal age

cbind(x,y) %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  full_join(., DGEList_placenta$samples, by = "samplename") %>% 
  ggplot() +
  geom_point(aes(x = x,
                 y = y,
                 colour = maternalAge,
                 shape = delivery_mode),
             size = 6) +
  geom_text(aes(x = x,
                y = y,
                label = samplename,
                colour = maternalAge), vjust = -1) +
  labs(title = "MDS plot filtered and normalised counts",
       x = "Dim 1",
       y = "Dim 2") +
  theme_bw()

## what is the relationship between gestational age at birth and birth weight?
ggplot(data = DGEList_placenta$samples,
       aes(x = gestationPlus,
                 y = birthwgt,
                 colour = fetal_sex)) +
  geom_point(aes(group = fetal_sex), size = 6) +
    labs(title = "Relationship between Gestational Age at birth and neonate weight at birth",
       x = "Gestational Age (Weeks')",
       y = "Birth Weight (grams)") +
  geom_smooth(aes(group = fetal_sex),
              se = FALSE, method = loess) +
  theme_bw(base_size = 14)

```

# Differential expression analysis

```{r differential Expression mode only}

DGEList_placenta$samples <- mutate(DGEList_placenta$samples, mode = ifelse(delivery_mode == "LSCS in labour", "LSCS_lab",
                                                    ifelse(delivery_mode == "Operative vaginal", "op_vag",
                                                           ifelse(delivery_mode == "Prelabout LSCS", "LSCS_pre",
                                                                  ifelse(delivery_mode == "Unassisted vaginal", "unassist_vag", NA)))))
                                   
                                   
levels(DGEList_placenta$samples$mode) <- c("LSCS_lab", "op_vag", "LSCS_pre", "unassist_vag")

# design to test DE between LSCS in labour and pre-labour LSCS
design_mode <- model.matrix(~0 + mode, 
                                   data = DGEList_placenta$samples)
colnames(design_mode) <- levels(DGEList_placenta$samples$mode)

voom_mode <- voomWithQualityWeights(DGEList_placenta, 
                                           design_mode, plot=TRUE)

fit_mode <- lmFit(voom_mode, design_mode)

fit_mode <- eBayes(fit_mode)

summary(decideTests(fit_mode)) # before FDR correction

contrasts <- c("LSCS_pre-LSCS_lab", "unassist_vag-op_vag")

contrast_matrix <- makeContrasts(contrasts = contrasts, levels = design_mode)

fit_contrast <- contrasts.fit(fit_mode, contrast_matrix)
fit_contrast <- eBayes(fit_contrast)
summ_fit_contrast <- summary(decideTests(fit_contrast, adjust.method="BH", p.value=0.05))

topTable <- topTable(fit_contrast, coef = 1)


```

## plot GOI



```{r differential Expression mode and fetal sex}

# design to test DE between LSCS in labour and pre-labour LSCS looking at female and male bearing pregnancies seperately
design_independent <- model.matrix(~ fetal_sex + 
                                     fetal_sex:delivery_mode, 
                                   data = DGEList_placenta$samples)

voom_independent <- voomWithQualityWeights(DGEList_placenta, 
                                           design_independent, plot=TRUE)

fit_independent <- lmFit(voom_independent, 
                         design_independent)

fit_independent <- eBayes(fit_independent)

summary(decideTests(fit_independent)) # before FDR correction

```
