---
title: "RNAseqAnalysis"
author: "Mhyles Hintural"
date: "26/08/2020"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set-up

```{r}
library(plyr)
library(reshape2)
library(dplyr)
library(tidyverse)
library(magrittr)
library(edgeR)
library(RColorBrewer)
library(limma)
library(ggplot2)
library(readxl)
library(pheatmap)
library(here)
library(xtable)
library(biomaRt)
library(org.Hs.eg.db)
library(factoextra)
```

# Data packaging
## Importing count data

```{r ImportCounts}
 
# Read in bam counts
termPlacentaCounts <- read.table(here("rawData/bam_counts.table.tsv"),
                                 sep = '\t',
                                 header = TRUE,
                                 row.names = "Geneid")
 
# clean up the column headers and remove the file extension names
names(termPlacentaCounts) <- gsub("_Aligned.sortedByCoord.out.bam", "", names(termPlacentaCounts))
 
 
names <- names(termPlacentaCounts) %>%
  as.data.frame() %>%
  separate(., col = ., into = c("cohort", "samplenumber", "tissue", "lane"), sep = "_") %>%
  mutate(., samplename = paste(cohort, samplenumber, sep = "")) %>%
  dplyr::select(., samplename)
 
# convert back to a vector so I can overwrite the colnames in the counts table
names <- names[,]
 
# replace the old sample names with the new names
names(termPlacentaCounts) <- names
```

## Importing meta-data

```{r ImportMetadata}

# Read in metadata

simpleMetadata <- read_excel(here("cleanData/simpleMetadata_matched.xlsx"))

```

## Creating a DGEList

```{r DGEList}

# new data frame of gene names for DGEList object
genes <- as.character(rownames(termPlacentaCounts)) %>%
  as.data.frame() %>%
  set_colnames("SYMBOL")

# new DGEList object for differential expression analysis
DGEList_placenta <- DGEList(counts = termPlacentaCounts,
                            samples = simpleMetadata,
                            genes = genes)

```

## Visualise distribution of expression prior to filtering

```{r}

cpm <- cpm(DGEList_placenta)
lcpm <-cpm(DGEList_placenta, log=TRUE)

nsamples <- ncol(DGEList_placenta)
col = colorRampPalette(brewer.pal(11,"Spectral"))(ncol(DGEList_placenta))
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.70), las=2,
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}

# Plot the distribution of un-normalised counts
un_normalised_lcpm <- cpm(DGEList_placenta, log=TRUE)
boxplot(un_normalised_lcpm, las=2, col=col, main="")
title(main = "Placenta: Un-normalised data (RNA)", ylab = "Log-cpm")

```

## Removing low expression genes

```{r filterLow}

# which genes are expressed at > 1 counts per million, in at least 6 samples
# we choose 6 samples because prelabour LSCS has 6 patients, which is our smallest group
keep.exprs <- rowSums(cpm >1) >= 6
keep.exprs %>% table()

# only keep the genes that met the criteria above
DGEList_placenta <- DGEList_placenta[keep.exprs,, keep.lib.sizes = FALSE]

```

## Visualise distribution of expression after filtering

```{r}

cpm <- cpm(DGEList_placenta)
lcpm_post <-cpm(DGEList_placenta, log=TRUE)

nsamples <- ncol(DGEList_placenta)
col = colorRampPalette(brewer.pal(11,"Spectral"))(ncol(DGEList_placenta))
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.70), las=2,
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm_post[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
# there seems to be one really big peak still, not sure why

# Plot the distribution of un-normalised counts
un_normalised_lcpm_post <- cpm(DGEList_placenta, log=TRUE)
boxplot(un_normalised_lcpm_post, las=2, col=col, main="")
title(main = "Placenta: Un-normalised data (RNA)", ylab = "Log-cpm")

```